src/com/goide/inspections/GoPlaceholderChecker.java

```diff
@@ -0,0 +1,497 @@
+    "fprint",
+    "fprintln",
+    "log",
+    "panic",
+    "panicln",
+    "print",
+    "println",
+    "sprint",
+    "sprintln"
+  );
+
+  private ProblemsHolder myHolder;
+  private GoCallExpr myCallExpr;
+  private GoFunctionOrMethodDeclaration myDeclaration;
+
+  public GoPlaceholderChecker(ProblemsHolder holder, GoCallExpr callExpr, GoFunctionOrMethodDeclaration declaration) {
+    myHolder = holder;
+    myCallExpr = callExpr;
+    myDeclaration = declaration;
+  }
+
+  // todo: implement ConstEvaluator
+  @Nullable
+  private static String getValue(@Nullable GoExpression expression) {
+    if (expression instanceof GoStringLiteral) {
+      return expression.getText();
+    }
+    if (expression instanceof GoAddExpr) {
+      String sum = getValue((GoAddExpr)expression);
+      return sum.isEmpty() ? null : sum;
+    }
+
+    return null;
+  }
+
+  // todo: implement ConstEvaluator
+  @NotNull
+  private static String getValue(@Nullable GoAddExpr expression) {
+    if (expression == null) return "";
+    StringBuilder result = new StringBuilder();
+    for (GoExpression expr : expression.getExpressionList()) {
+      if (expr instanceof GoStringLiteral) {
+        result.append(expr.getText());
+      }
+      else if (expr instanceof GoAddExpr) {
+        result.append(getValue(expr));
+      }
+    }
+
+    return result.toString();
+  }
+
+  private static int getPlaceholderPosition(@NotNull GoFunctionOrMethodDeclaration function) {
+    Integer position = FORMATTING_FUNCTIONS.get(StringUtil.toLowerCase(function.getName()));
+    if (position != null) {
+      String importPath = function.getContainingFile().getImportPath(false);
+      if ("fmt".equals(importPath) || "log".equals(importPath) || TESTING_PATH.equals(importPath)) {
+        return position;
+      }
+    }
+    return -1;
+  }
+
+  @Nullable
+  private static String resolve(@NotNull GoExpression argument) {
+    if (argument instanceof GoStringLiteral) return argument.getText();
+
+    PsiReference reference = argument.getReference();
+    PsiElement resolved = reference != null ? reference.resolve() : null;
+
+    if (resolved instanceof GoVarDefinition) {
+      return getValue(((GoVarDefinition)resolved).getValue());
+    }
+    if (resolved instanceof GoConstDefinition) {
+      return getValue(((GoConstDefinition)resolved).getValue());
+    }
+
+    return null;
+  }
+
+  public void checkPrint() {
+    // TODO florin: Implement the print function checking
+    // Ref code: https://github.com/golang/go/blob/79f7ccf2c3931745aeb97c5c985b6ac7b44befb4/src/cmd/vet/print.go#L587
+  }
+
+  private class FormatState {
+    public char verb;             // the format verb: 'd' for "%d"
+    public String format;         // the full format directive from % through verb, "%.3d".
+    public String flags;          // the list of # + etc.
+    public List<Integer> argNums; // the successive argument numbers that are consumed, adjusted to refer to actual arg in call
```

Comment:
it's always initialized in `parsePrintfVerb` and it's always `new ArrayList`. Do you have any reason to not to declare it as final